server/src/main/java/org/opensearch/common/logging/DeprecatedMessage.java

```diff
@@ -77,6 +82,34 @@ private static Map<String, Object> fieldMap(String key, String xOpaqueId) {
     }
 
     public boolean isAlreadyLogged() {
-        return !keys.add(keyWithXOpaqueId);
+        return !keyDedupeCache.add(keyWithXOpaqueId);
+    }
+
+    private static class KeyDedupeCache {
+        private static final Object VALUE = new Object();
+        private final Cache<Integer, Object> cache;
+
+        private KeyDedupeCache() {
+            cache = CacheBuilder.<Integer, Object>builder()
+                .weigher((s, o) -> 1) // Entries are a fixed size (integers)
+                .setMaximumWeight(MAX_DEDUPE_CACHE_ENTRIES)
+                .build();
+        }
+
+        boolean add(String key) {
+            int hash = key.hashCode();
```

Comment:
DeprecationLogger is called from lot of different code paths including every search request. The X-Opaque-Id length is not controlled by the OpenSearch process and generating hashCode can be expensive for really long strings. Though I agree storing hashCode as key would limit the memory usage but would add compute overhead.

Reply:
hashcode() is going to be called no matter what if we put the string in a set/map, right?